File: app/src/main/java/com/attendance/courses/GetCoursesSubscriber.java
package com.attendance.courses;

import com.attendance.AttendanceApplication;
import com.attendance.R;

import java.net.ConnectException;
import java.net.SocketTimeoutException;

import rx.Subscriber;

/**
 * Created by peiqin on 3/5/2017.
 */

public class GetCoursesSubscriber extends Subscriber<Boolean> {

    private CoursesContract.View view;

    public GetCoursesSubscriber(CoursesContract.View view) {
        this.view = view;
    }

    @Override
    public void onStart() {
        view.startRefresh();
    }

    @Override
    public void onCompleted() {
        view.stopRefresh();
        if (!this.isUnsubscribed()) {
            this.unsubscribe();
        }
    }

    @Override
    public void onError(Throwable e) {

        if (e instanceof SocketTimeoutException || e instanceof ConnectException) {
            view.showTip(AttendanceApplication.getContext().getString(R.string.network_corrupt));
        } else {
            view.showTip("error:" + e.getMessage());
        }
        view.stopRefresh();

        if (!this.isUnsubscribed()) {
            this.unsubscribe();
        }

    }

    @Override
    public void onNext(Boolean result) {

        if (result) {
            view.getCourseSuccess();
            view.showTip(AttendanceApplication.getContext().getString(R.string.get_success));
        } else {
            view.showTip(AttendanceApplication.getContext().getString(R.string.get_failed));
        }

    }

}

File: app/src/main/kotlin/com.attendance/courses/GetCoursesSubscriber.kt
package com.attendance.courses

import com.attendance.AttendanceApplication
import com.attendance.R
import com.attendance.courses.CoursesContract

import java.net.ConnectException
import java.net.SocketTimeoutException

import rx.Subscriber

/**
 * Created by peiqin on 3/5/2017.
 */

class GetCoursesSubscriber(private val view: CoursesContract.View) : Subscriber<Boolean>() {

    override fun onStart() {
        view.startRefresh()
    }

    override fun onCompleted() {
        view.stopRefresh()
        if (!this.isUnsubscribed) {
            this.unsubscribe()
        }
    }

    override fun onError(e: Throwable) {

        if (e is SocketTimeoutException || e is ConnectException) {
            view.showTip(AttendanceApplication.context.getString(R.string.network_corrupt))
        } else {
            view.showTip("error:" + e.message)
        }
        view.stopRefresh()

        if (!this.isUnsubscribed) {
            this.unsubscribe()
        }

    }

    override fun onNext(result: Boolean?) {

        if (result!!) {
            view.getCourseSuccess()
            view.showTip(AttendanceApplication.context.getString(R.string.get_success))
        } else {
            view.showTip(AttendanceApplication.context.getString(R.string.get_failed))
        }

    }

}
